Fix: Cronjob will no longer works when updated in sometimes

yanhongchang opened the pull request:
when update cronjob after posting a job and before update the status, the cronjob status will update failed, but the job has been created, the next sync time, will create again and get a job without nothing, then the nil job will update into the cronjob status active list.  
for example：

                ...
		status:
		 active:
		 - apiVersion: batch/v1
		   kind: Job
		lastScheduleTime: "2022-08-17T02:05:00Z"

the cronjob will do not work any more.

/kind bug

yanhongchang commented:
/ok-to-test

yanhongchang commented:
@soltysh @kow3ns Hi, please review my pr, thanks!

pacoxu commented:
/ok-to-test

pacoxu commented:
/cc @alaypatel07

yanhongchang commented:
retest

yanhongchang commented:
/retest

yanhongchang commented:
/test

yanhongchang commented:
/test all

yanhongchang commented:
/test

yanhongchang commented:
> ```
> /retest
> ```
> 
> will work if some test ci failed.

OK, thanks!

yanhongchang commented:
@pacoxu Hi pacoxu, I am a little confused，Why is there no reviewer to review the pr ? Is it to take a lot of time or something wrong in my pr?

pacoxu commented:
> @pacoxu Hi pacoxu, I am a little confused，Why is there no reviewer to review the pr ? Is it to take a lot of time or something wrong in my pr?

To quote some answers when I ask similar questions.
> This goes slow, reviewers are busy people and there are priorities
> You can attend sig weekly or bi-weekly meetings, and I think they do a bug triage
BTW, for the [v1.26 release](https://www.kubernetes.dev/resources/release/#timeline) cycle, PRR freeze and enhancement freeze is coming and many reviewers/approvers are busy on that.

pacoxu commented on pkg/controller/cronjob/cronjob_controllerv2_test.go:
```diff
@@ -462,14 +462,15 @@ func TestControllerV2SyncCronJob(t *testing.T) {
 			jobPresentInCJActiveStatus: true,
 		},
-		"prev ran but done, is time, create job failed, A": {
+		"prev ran but done, is time, job has existed, but get failed, A": {
```
This test case was added in #107470.
/cc @jlsong01 @alculquicondor
It seems to be expected behavior.

pacoxu replied:
This test case was added in #107470.
/cc @jlsong01 @alculquicondor
It seems to be expected behavior.

alculquicondor commented:
```golang
// If the job is created by other actor, assume it has updated the cronjob status accordingly
```

This line makes me think that this was intended. @soltysh ?

yanhongchang commented:
@soltysh please attention, thanks~

yanhongchang commented:
@pacoxu Hi, Why do these reviewers not reply for so long？Is really need so much time?

pacoxu commented:
> ```go
> // If the job is created by other actors, assume it has updated the cronjob status accordingly
> ```
> 
> This line makes me think that this was intended. @soltysh ?

kindly ping  @jlsong01 and @soltysh, do you have some input on this?



> @pacoxu Hi, Why do these reviewers not reply for so long？ Is really need so much time?

If this pr/issue confused me a lot, I may ask in #sig-app slack or sig-app meeting for some suggestions.

BTW, is there a behavior change after #107470? Or is this different behavior from cronjob controller v1?

soltysh commented on pkg/controller/cronjob/cronjob_controllerv2.go:
```diff
@@ -596,22 +596,32 @@ func (jm *ControllerV2) syncCronJob(
+
+		// When the job created by other actor and the status of cronjob has not been
+		// updated normally, so here need get all jobs again.
+		jobResp, err = jm.jobControl.GetJob(cronJob.GetNamespace(), jobReq.GetName())
```
If you want to tackle this case, which looks like a valid one based on your description, I'd suggest the following steps to solve it:
- in this section, when you receive `AlreadyExists` error, you'd need to refresh the object (just like you do here), but additionally you'd need to check owner reference to ensure that the controller is the one who created it. I don't envision us adopting orphaned jobs, that's why the owner ref check is critical for this to work. 

Additionally, I'd like to see a test, which confirms this case is properly covered in both cases:
- pre-existing job with matching owner ref (your fix)
- pre-existing job with non-matching owner ref (current path).

yanhongchang replied:
Ok, I have added the logic part of check the owner.

soltysh reviewed (changes requested):
Please ping me on slack, once you have this fixed per my comment.

pacoxu commented:
/remove-lifecycle stale

yanhongchang committed:
Fix: Cronjob will no longer work when updated in sometimes

when update cronjob after posting a job and before update the status,
the cronjob status will update failed, but the job has been created,
the next sync time, will create again and get a job without nothing,
then the nil job will update into the cronjob status active list.
the cronjob will do not work any more.

yanhongchang commented:
/retest-required
